### DISCUSSION TITLE
InputBox

### DISCUSSION THREAD

**Discussion by jfoshee** [tui-cs/Terminal.Gui#2089] [Ideas]
Would an `InputBox` be a desirable addition?  I was looking for one myself.

Here is a start, largely derived from UICatalog `TableEditor.EditCurrentCell`.

```cs
public static class InputBox
{
    public static string? Show(string prompt, string? title, string? defaultValue)
    {
        bool okPressed = false;

        var ok = new Button("Ok", is_default: true);
        ok.Clicked += () => { okPressed = true; Application.RequestStop(); };
        var cancel = new Button("Cancel");
        cancel.Clicked += () => { Application.RequestStop(); };
        var dialog = new Dialog(title, ok, cancel);

        var label = new Label(prompt)
        {
            X = 0,
            Y = 1
        };

        var textField = new TextField(defaultValue ?? "")
        {
            X = 0,
            Y = 2,
            Width = Dim.Fill()
        };

        dialog.Add(label, textField);
        textField.SetFocus();

        Application.Run(dialog);

        if (okPressed)
        {
            return textField.Text.ToString();
        }
        return null;
    }
}
```

**Comment by tig** [comment]
Nice idea. 

I wonder if this could be more generic though, and support any `View` that has a `Text` property? 

```csharp
var tree = new TreeView() { ... };
...
var result = InputBox.Query("Select from Tree", tree);

MessageBox.Query("You selected..", result);

// Above is same as:
MessageBox.Query("You selected..", tree.Text);
```

(I think the `Label` is superfluous given the `Dialog` has a `Title`).

**Comment by tznind** [comment]
I think a basic InputBox would be nice to have, it is handy and would lower the bar of entry.   Given we have `MessageBox` and `OpenDialog` it makes sense to have an easy way of getting input from user too.

I think that we **should** support (optional) Label text.  The traditional input box supports both title and label and they have different roles.  

- Title gives the user context for what is expected
- Label gives detail on what they are supposed to enter.  This text could get quite long and may even require wrapping.

![image](https://user-images.githubusercontent.com/31306100/196110797-5f01af3b-0516-43e9-a789-e6f183ca8a30.png)

https://learn.microsoft.com/en-us/office/vba/language/reference/user-interface-help/inputbox-function

https://www.w3schools.com/js/js_popup.asp

I love the idea of making it slottable.  Building on that idea, I think we should start with only public static method signature that matches other libraries (e.g. VB / Javascript).  We can keep the 'slottable' backend private until we are sure the code design is finalised and then make public if there is benefit from it.

Something like this:

```csharp
public class InputBox<T> : Dialog
{
    private readonly Func<T> getter;

    public T Result { get
        {
            return getter();
        }
    }

    private InputBox(string title, string prompt, T defaultValue, View hosted, Func<T> getter)
    {
        this.getter = getter;
        // TODO setup rest of view
    }

    public static string QueryString(string title, string prompt, string defaultValue)
    {
        var tf = new TextField();
        var getter = ()=>tf.Text.ToString();
        var box = new InputBox<string>(title, prompt, defaultValue, tf, getter);
        Application.Run(box);
        return box.Result;

    }
    public static DateTime QueryDate(string title, string prompt, DateTime defaultValue)
    {
        var df = new DateField();
        var getter = () => df.Date;
        var box = new InputBox<DateTime>(title, prompt, defaultValue, df, getter);
        Application.Run(box);
        return box.Result;
    }
}
```
